Logging rewiring - derive from non-default root logger

Helcaraxan opened the pull request:
After #662 this is the second (and last) PR for the logger rewiring.

Instead of relying on the standard logrus logger instance and tweaking it at various places we will now derive from a single non-default instance that we instantiate in our `main()` (except for the Matterclient which now gets its own).

42wim commented:
@Helcaraxan sorry for the hound spam, wasn't suppose to do that. Disabled it for now.
But the build has failed because of extra vendoring needed.

Helcaraxan commented:
Was half-expecting so. Have fixed the missing vendoring. Don't worry about the Hound spam. Was not expecting it but it does encourage you to Do The Right Thing™️.

42wim commented on gateway/gateway.go:
```diff
@@ -31,25 +33,30 @@ type BrMsgID struct {
+// New creates a new Gateway object associated with the specified router and
+// following the given configuration.
+func New(rootLogger *logrus.Logger, cfg config.Gateway, r *Router) *Gateway {
+	logger := logrus.WithFields(logrus.Fields{"prefix": "gateway"})
```
```suggestion
        logger := rootLogger.WithFields(logrus.Fields{"prefix": "gateway"})
```

42wim commented on gateway/gateway.go:
```diff
@@ -69,15 +76,18 @@ func (gw *Gateway) FindCanonicalMsgID(protocol string, mID string) string {
-		// set logging
-		br.Log = logrus.WithFields(logrus.Fields{"prefix": "bridge"})
-		brconfig := &bridge.Config{Remote: gw.Message, Log: logrus.WithFields(logrus.Fields{"prefix": br.Protocol}), Bridge: br}
+		br.Log = logrus.WithFields(logrus.Fields{"prefix": br.Protocol})
```
```suggestion
                br.Log = gw.logger.WithFields(logrus.Fields{"prefix": br.Protocol})
```

42wim commented on gateway/router.go:
```diff
@@ -16,15 +17,22 @@ type Router struct {
+// NewRouter initializes a new Matterbridge router for the specified configuration and
+// sets up all required gateways.
+func NewRouter(rootLogger *logrus.Logger, cfg config.Config, bridgeMap map[string]bridge.Factory) (*Router, error) {
+	logger := rootLogger.WithField("prefix", "router")
```
```suggestion
	logger := rootLogger.WithFields(logrus.Fields{"prefix": "router"})
```

Helcaraxan replied:
That's really weird. That should not make a difference. 😱

42wim commented:
Testing it now, have add some changes to get the same output as before.

42wim commented on matterbridge.go:
```diff
@@ -10,53 +10,76 @@ import (
+			ForceFormatting: true,
+		}
+		logger.Level = logrus.DebugLevel
+		logger.Info("Enabling debug logging.")
```
```suggestion
		logger.WithFields(logrus.Fields{"prefix": "main"}).Info("Enabling debug logging.")
```

42wim commented on matterbridge.go:
```diff
@@ -10,53 +10,76 @@ import (
 	"github.com/42wim/matterbridge/gateway/bridgemap"
 	"github.com/google/gops/agent"
-	prefixed "github.com/matterbridge/logrus-prefixed-formatter"
+	"github.com/matterbridge/logrus-prefixed-formatter"
```
```suggestion
	prefixed "github.com/matterbridge/logrus-prefixed-formatter"
```
because the path isn't the same as the package name, makes it better visible (unless you have strong arguments to not do this)

Helcaraxan replied:
Yeah my linter was actually complaining about it. Have reinstated it.

42wim commented on matterclient/matterclient.go:
```diff
@@ -10,7 +10,7 @@ import (
 	"github.com/hashicorp/golang-lru"
 	"github.com/jpillora/backoff"
-	prefixed "github.com/matterbridge/logrus-prefixed-formatter"
+	"github.com/matterbridge/logrus-prefixed-formatter"
```
```suggestion
	prefixed "github.com/matterbridge/logrus-prefixed-formatter"
```

42wim commented:
Testing done, looks good 👍 
Thanks! If you take care of my remarks, it can be merged.

Helcaraxan commented:
@42wim there is relatively large diff with the previous versions as I stumbled upon other linter errors pertaining to some big structs that were being passed by value. Please review with care. You can use the [force-push link](https://github.com/42wim/matterbridge/compare/3336747afbcba416bcc4b436ee05be7104061c45..7dbaf596017a7c2781cb80804b6763eda1cead21) to see the exact diff between the previous version and the current one.

42wim commented:
Ok, maybe next time better split up in multiple PR's or multiple commits as this is better for when we have to look up why we exactly did those changes :-)

42wim commented on matterclient/matterclient.go:
```diff
@@ -189,46 +223,48 @@ func (m *MMClient) WsReceiver() {
-					}
-					if m.OnWsConnect != nil {
-						m.OnWsConnect()
+						m.logger.Errorf("Login failed: %#v", err)
```
Could you revert everything not logger related in this function for now?

42wim commented:
@Helcaraxan do you have some time to look into my last comment? I'd really like to merge this :-)

Helcaraxan commented:
@42wim sorry for the near month without progress on this. I've just rebased on top of the current `HEAD` and addressed your comment about `StatusLoop` in `matterclient/matterclient.go`.

42wim commented:
Seems to crash each time here:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x638b98]

goroutine 553 [running]:
github.com/sirupsen/logrus.(*Entry).Log(0x0, 0xc000000005, 0xc0006c1c90, 0x1, 0x1)
        /wim/go/pkg/mod/github.com/sirupsen/logrus@v1.3.0/entry.go:255 +0x48
github.com/sirupsen/logrus.(*Entry).Logf(0x0, 0xc000000005, 0x148ef4a, 0x16, 0xc000574ef8, 0x1, 0x1)
        /wim/go/pkg/mod/github.com/sirupsen/logrus@v1.3.0/entry.go:301 +0xed
github.com/sirupsen/logrus.(*Entry).Debugf(0x0, 0x148ef4a, 0x16, 0xc000574ef8, 0x1, 0x1)
        /wim/go/pkg/mod/github.com/sirupsen/logrus@v1.3.0/entry.go:309 +0x73
github.com/42wim/matterbridge/bridge/slack.(*Bslack).handleSlackClient(0xc000392ea0, 0xc0000c72c0)
        /wim/go/src/github.com/42wim/matterbridge/bridge/slack/handlers.go:47 +0x114b
created by github.com/42wim/matterbridge/bridge/slack.(*Bslack).handleSlack
        /wim/go/src/github.com/42wim/matterbridge/bridge/slack/handlers.go:20 +0xbba
```

42wim commented:
Thanks a lot 👍  
Got merged via #728

KrzysztofMadejski commented:
@Helcaraxan I have been working on a WhatsApp bridge in parallel, should I update there logging to keep it in line with your solution? 

https://github.com/42wim/matterbridge/pull/711/files

Helcaraxan commented:
@KrzysztofMadejski yes that would be best indeed. If you want me to take a stab at it feel free to ask. 🙂

KrzysztofMadejski commented:
Either you can tell what to do, or you can take a stab at it. Whatever you feel will be faster / more convenient for you. :)
